File: domain/src/main/java/com/study/platform/domain/comment/model/Comment.java
package com.study.platform.domain.comment.model;

import com.study.platform.domain.comment.event.CommentCreatedEvent;
import com.study.platform.domain.post.model.StudyPost;
import com.study.platform.domain.user.model.User;
import com.study.platform.global.entity.BaseTimeEntity;
import com.study.platform.global.event.DomainEventPublisher;
import com.study.platform.global.exception.CustomException;
import com.study.platform.global.exception.ErrorCode;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.UUID;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(
        name = "comments",
        indexes = {
                @Index(name = "idx_comment_post_created", columnList = "post_id, created_at")
        }
)
public class Comment extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "BINARY(16)")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id", nullable = false)
    private StudyPost post;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String content;

    @Builder
    private Comment(StudyPost post, User author, String content) {
        this.post = post;
        this.author = author;
        this.content = content;
    }

    public static Comment create(StudyPost post, User author, String content, DomainEventPublisher publisher) {
        Comment comment = Comment.builder()
                .post(post)
                .author(author)
                .content(content)
                .build();
        if (post != null && !post.isAuthor(author.getId())) {
            publisher.publish(new CommentCreatedEvent(
                    post.getId(), post.getAuthor().getId(), author.getId(), post.getTitle()));
        }
        return comment;
    }

    public void update(String content) {
        this.content = content;
    }

    public boolean isAuthor(UUID userId) {
        return this.author.getId().equals(userId);
    }

    public void validateAuthor(UUID userId) {
        if (!isAuthor(userId)) {
            throw new CustomException(ErrorCode.NOT_COMMENT_AUTHOR);
        }
    }
}

File: domain/src/main/kotlin/com/study/platform/domain/comment/model/Comment.kt
package com.study.platform.domain.comment.model

import com.study.platform.domain.comment.event.CommentCreatedEvent
import com.study.platform.domain.post.model.StudyPost
import com.study.platform.domain.user.model.User
import com.study.platform.global.entity.BaseTimeEntity
import com.study.platform.global.event.DomainEventPublisher
import com.study.platform.global.exception.CustomException
import com.study.platform.global.exception.ErrorCode
import jakarta.persistence.*
import java.util.UUID

@Entity
@Table(
    name = "comments",
    indexes = [Index(name = "idx_comment_post_created", columnList = "post_id, created_at")]
)
class Comment : BaseTimeEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "BINARY(16)")
    var id: UUID? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id", nullable = false)
    var post: StudyPost? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    var author: User? = null

    @Column(columnDefinition = "TEXT", nullable = false)
    var content: String = ""

    fun update(content: String) {
        validateContent(content)
        this.content = content
    }

    fun isAuthor(userId: UUID): Boolean = author?.id == userId

    fun validateAuthor(userId: UUID) {
        if (!isAuthor(userId)) throw CustomException(ErrorCode.NOT_COMMENT_AUTHOR)
    }

    companion object {
        private const val MAX_CONTENT_LENGTH = 500

        fun create(post: StudyPost, author: User, content: String, publisher: DomainEventPublisher): Comment {
            validateContent(content)
            val comment = Comment().also {
                it.post = post
                it.author = author
                it.content = content
            }
            if (!post.isAuthor(author.id!!)) {
                publisher.publish(CommentCreatedEvent(post.id!!, post.author!!.id!!, author.id!!, post.title))
            }
            return comment
        }

        private fun validateContent(content: String) {
            if (content.isBlank()) throw CustomException(ErrorCode.INVALID_INPUT)
            if (content.length > MAX_CONTENT_LENGTH) throw CustomException(ErrorCode.INVALID_INPUT)
        }
    }
}
